### PR TITLE
Display Custom Fields on ContributionPage settings

### PR DESCRIPTION
**Pull request by aydun**
Overview
----------------------------------------
This adds the display of Custom Fields on ContributionPages. 

Before
----------------------------------------
Custom Fields on ContributionPages can be configured by API but not displayed.

After
----------------------------------------
Custom Fields on ContributionPages are displayed at the bottom of the Settings tab.

image

Technical Details
----------------------------------------

Comments
----------------------------------------
The customDataBlock.tpl that pulls in custom fields only loads groups with `style` of `'Inline'`.  

The GDPR extension creates custom fields on ContributionPages with `Inline` style but has its own code to display these as a tab.  This PR causes those fields to also be shown on the Settings tab.  This can be avoided by changing the style to `Tab` - I will create a PR for GDPR.

### CODE REVIEW

**Comment by demeritcowboy** [comment]
Looks good. Putting merge-ready just because it's not clear if it's intentional/out-of-scope that it doesn't work with subtypes, e.g. a custom group that is only for ContributionPages with financial type Donation. You can define such a custom group, but then it doesn't display here. I think it's ok for this if it's out-of-scope.

Test fails are also happening elsewhere.

**Comment by seamuslee001** [comment]
Whilst I agree with this work, I think that we may want to think about some kind of upgrade step for people with this extension https://lab.civicrm.org/extensions/contributionpagecustomfield

**Comment by seamuslee001** [comment]
Just to clarify the extension I was referring to above currently handles the ability to have custom fields on contribution pages and also displays them. So if we merge this then people upgrading would probably end up possibly with the custom data displaying twice right? So I was thinking we could essentially pull this option value into core https://lab.civicrm.org/extensions/contributionpagecustomfield/-/blob/master/contributionpagecustomfield.php?ref_type=heads#L42 and then effectively deprecate that whole extension at that point and have an upgrade message of like "We see that you had the Contribution Page Custom Field extension enabled, this has now been disabled as it is now incorporated in core or similar"

**Comment by colemanw** [comment]
@seamuslee001 actually that wheel has already been invented :)
Just push an intro into this file (could do it as part of this PR for completeness): https://github.com/civicrm/civicrm-core/blob/master/extension-compatibility.json

**Comment by aydun** [comment]
> Looks good. Putting merge-ready just because it's not clear if it's intentional/out-of-scope that it doesn't work with subtypes, e.g. a custom group that is only for ContributionPages with financial type Donation. You can define such a custom group, but then it doesn't display here. I think it's ok for this if it's out-of-scope.
> 
> Test fails are also happening elsewhere.

@demeritcowboy Thanks - It's kinda out of scope in that I didn't know you could do that and don't need it.  But I'm happy to try putting that in for completeness.  However, the Custom Groups UI does not provide that option and I don't see how to do it with APIv4 Explorer.  Could you provide more detail please?

**Comment by demeritcowboy** [comment]
It's `option_value.grouping`. See https://docs.civicrm.org/dev/en/latest/step-by-step/create-entity/#121-making-our-entity-available-for-custom-data

And then the {include} in the tpl has a parameter customDataSubType. I don't know how magic it is - if it can handle anything you throw at it.

**Comment by aydun** [comment]
@demeritcowboy - I think that now handles financial_type subtypes.

**Comment by aydun** [comment]
@seamuslee001 @colemanw I've enabled custom fields on ContributionPage in core, and updated extension-compatibility.json to obsolete the contributionpagecustomfield extension.

FYI, if you're testing [this](https://lab.civicrm.org/extensions/contributionpagecustomfield/-/merge_requests/8) needs to be applied to make the contributionpagecustomfield extension work on master (prior to this PR).

I think that covers everything ...

File: templates/CRM/Contribute/Form/ContributionPage/Settings.tpl [hunk 1]
```diff
@@ -10,158 +10,170 @@
...
+  {include file="CRM/common/customDataBlock.tpl" customDataType='ContributionPage' customDataSubType=$financialTypeId cid=''}
+  <div class="crm-submit-buttons">{include file="CRM/common/formButtons.tpl" location="bottom"}</div>
 </div>
 
+<script type="text/javascript">
+  CRM.$(function($) {
+    $('#financial_type_id').change( function() {
+      $('#customData_ContributionPage > *').remove();
+      $('#customData_ContributionPage').append(
+        CRM.buildCustomData('ContributionPage', $('#financial_type_id').val(), false, false, false, false, false, false)
```
Comment (demeritcowboy) [inline templates/CRM/Contribute/Form/ContributionPage/Settings.tpl:143]: The remove() and append() don't seem needed and I haven't seen those elsewhere. It works with just the buildCustomData line. It uses magic. Dark magic.

**Comment by demeritcowboy** [comment]
Looks pretty good. One thing missing for parity with event custom fields, but isn't a blocker for this PR, is that for events this setting:

untitled3

controls whether it appears on the public Event Info page. It also says it controls whether it appears in receipts but that seems less important.

untitled4

 Whether this is desirable for contribution pages is a different question, but if not, then that setting could be confusing.

File: templates/CRM/Contribute/Form/ContributionPage/Settings.tpl (hunk 1 above)
Comment (aydun) [reply templates/CRM/Contribute/Form/ContributionPage/Settings.tpl]: Even more magic than I thought!  Updated now.

**Comment by demeritcowboy** [comment]
Sold!

**Comment by aydun** [comment]
Many thanks for your help with this PR @demeritcowboy

**Pull request merged**